### ISSUE TITLE
Conflict between require_ssl and ssl_mode settings for google_sql_database_instance

### ISSUE THREAD

**Issue by gygitlab** [hashicorp/terraform-provider-google#17443]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to a user, that user is claiming responsibility for the issue.
* Customers working with a Google Technical Account Manager or Customer Engineer can ask them to [reach out internally](https://github.com/hashicorp/terraform-provider-google/wiki/Customer-Contact#raising-gcp-internal-issues-with-the-provider-development-team) to expedite investigation and resolution of this issue.

### Terraform Version

1.7.4

### Affected Resource(s)

`google_sql_database_instance`

### Terraform Configuration

```tf
resource "google_sql_database_instance" "postgres_instance" {
  (...)
  settings {
    tier = "db-custom-2-7680"
    ip_configuration {
      # this should be value of "ALLOW_UNENCRYPTED_AND_ENCRYPTED", "ENCRYPTED_ONLY" or "TRUSTED_CLIENT_CERTIFICATE_REQUIRED"

      sslMode = "TRUSTED_CLIENT_CERTIFICATE_REQUIRED"
    }
  }
}
```

### Expected Behavior

That you can set `ssl_mode` cleanly for Cloud SQL databases without needing to also manage the setting it's replaced - The legacy `require_ssl` setting.

Just wanted to feed back on this one. [A PR was completed last year ](https://github.com/GoogleCloudPlatform/magic-modules/pull/9367) but it's a bit confusing for PostgreSQL as the guidance in the [Terraform docs](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/sql_database_instance#ssl_mode) appears conflict with the [GCP docs](https://cloud.google.com/sql/docs/postgres/admin-api/rest/v1/instances).

Specifically it's clear that `requireSsl` is now "legacy" but since it appears the Terraform provider sets it to `false` by default it now creates a conflict when trying to set `sslMode` to `TRUSTED_CLIENT_CERTIFICATE_REQUIRED` as the GCP docs states:

> Only allow connections encrypted with SSL/TLS and with valid client certificates. When this value is used, the legacy requireSsl flag **must be true or cleared** to avoid the conflict between values of two flags.

IMO for this issue to be completed it would seem `requireSsl` is deprecated / removed for PostgreSQL & MySQL instances where it's default is now "cleared" instead of `false` to allow for a clean replacement with `sslMode` because as it stands today you need to confusingly manage two separate variables here to avoid conflicts with setting SSL.

### Actual Behavior

The [`ssl_mode` variable](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/sql_database_instance#ssl_mode) was introduced for the `google_sql_database_instance` resource [last year](https://github.com/GoogleCloudPlatform/magic-modules/pull/9367).

This variable adds more options for Cloud SQL SSL, specifically allowing for 1 way enforcement, and effectively replaces the now "legacy" [`require_ssl` boolean](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/sql_database_instance#require_ssl).

However it looks like there is a conflict with this transition. The provider looks to set `require_ssl` to `false` when the user has not configured it. This creates a conflict with the [GCP API](https://cloud.google.com/sql/docs/postgres/admin-api/rest/v1/instances) that is basically expecting this to be "cleared" when the superseding `ssl_mode` is used or to be set specifically for backwards compatibility reasons.

Since `require_ssl` is always `false` instead of cleared this means users need to manage both settings in confusing fashion as the [API needs `require_ssl` to be set differently *if it is present* depending on what `ssl_mode`  is set to](https://cloud.google.com/sql/docs/postgres/admin-api/rest/v1/instances#sslmode). This adds brittleness in Terraform config.

TL;DR When setting `ssl_mode` to `TRUSTED_CLIENT_CERTIFICATE_REQUIRED` it will fail as the provider is incorrectly setting `require_ssl` to `false` in the background even if the user is not setting it at all:

```
│ Error: expected settings.0.ip_configuration.0.ssl_mode to be one of ["ALLOW_UNENCRYPTED_AND_ENCRYPTED" "ENCRYPTED_ONLY" "TRUSTED_CLIENT_CERTIFICATE_REQUIRED"], got
│
│   on cloud_sql.tf line 28, in resource "google_sql_database_instance" "sql":
│   28:       ssl_mode        = "TRUSTED_CLIENT_CERTIFICATE_REQUIRED"
```

Additional: [The Terraform documentation](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/sql_database_instance#ssl_mode) also contains several spelling errors.

### Steps to reproduce

1. Configure a `google_sql_database_instance` for PostgreSQL with it's `ssl_mode` set to `TRUSTED_CLIENT_CERTIFICATE_REQUIRED`.
2. Noticed it fails as it also "needs" `require_ssl` to be set but only because the provider is setting it to `false` in the first place.

### References

- https://github.com/hashicorp/terraform-provider-google/issues/16406
- https://github.com/GoogleCloudPlatform/magic-modules/pull/9367

b/327482344

**Comment by jt185289** [comment]
would love to note that this is just as confusing and convoluted if you want to use "ENCRYPTION_ONLY"
setting 

had set my terraform to:
  require_ssl = false
  ssl_mode    = "ENCRYPTED_ONLY"
  
  and that didnt provide the intended results as the GCP docs suggest and set our config to 'TRUSTED_CLIENT_CERTIFICATE_REQUIRED' instead.

**Comment by c2thorn** [comment]
https://github.com/GoogleCloudPlatform/magic-modules/pull/10797
and https://github.com/GoogleCloudPlatform/magic-modules/pull/10833 related

**Comment by c2thorn** [comment]
closed with https://github.com/GoogleCloudPlatform/magic-modules/pull/11154
